In [73]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

In [74]:
df = pd.read_csv('course_lead_scoring.csv')
df.head()

lead_source    industry  number_of_courses_viewed  annual_income  \
0      paid_ads         NaN                         1        79450.0   
1  social_media      retail                         1        46992.0   
2        events  healthcare                         5        78796.0   
3      paid_ads      retail                         2        83843.0   
4      referral   education                         3        85012.0   

  employment_status       location  interaction_count  lead_score  converted  
0        unemployed  south_america                  4        0.94          1  
1          employed  south_america                  1        0.80          0  
2        unemployed      australia                  3        0.69          1  
3               NaN      australia                  1        0.87          0  
4     self_employed         europe                  3        0.62          1

In [75]:
df.columns

Index(['lead_source', 'industry', 'number_of_courses_viewed', 'annual_income',
       'employment_status', 'location', 'interaction_count', 'lead_score',
       'converted'],
      dtype='object')

In [76]:
df.converted.unique()

array([1, 0])

In [77]:
df.lead_source.unique()

array(['paid_ads', 'social_media', 'events', 'referral', 'organic_search',
       nan], dtype=object)

In [78]:
df.industry = df.industry.fillna('NA')
df.employment_status = df.employment_status.fillna('NA')
df.location = df.location.fillna('NA')
df.lead_source = df.lead_source.fillna('NA')

### Question 1

What is the most frequent observation (mode) for the column `industry`?

- `NA`
- `technology`
- `healthcare`
- `retail`

In [79]:
df.industry.value_counts()

industry
retail           203
finance          200
other            198
healthcare       187
education        187
technology       179
manufacturing    174
NA               134
Name: count, dtype: int64

### Question 2

Create the [correlation matrix](https://www.google.com/search?q=correlation+matrix) for the numerical features of your dataset.
In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

- `interaction_count` and `lead_score`
- `number_of_courses_viewed` and `lead_score`
- `number_of_courses_viewed` and `interaction_count`
- `annual_income` and `interaction_count`

Only consider the pairs above when answering this question.

In [80]:
numerical = ['annual_income', 'number_of_courses_viewed', 'interaction_count', 'lead_score']
categorical = ['lead_source','industry', 'employment_status', 'location']

In [81]:
df[numerical].corrwith(df.converted)

annual_income               0.078256
number_of_courses_viewed    0.435914
interaction_count           0.374573
lead_score                  0.193673
dtype: float64

In [82]:
df[['interaction_count', 'lead_score']].corrwith(df.converted)

interaction_count    0.374573
lead_score           0.193673
dtype: float64

In [83]:
df[['number_of_courses_viewed', 'lead_score']].corrwith(df.converted)

number_of_courses_viewed    0.435914
lead_score                  0.193673
dtype: float64

In [84]:
df[['number_of_courses_viewed', 'interaction_count']].corrwith(df.converted)

number_of_courses_viewed    0.435914
interaction_count           0.374573
dtype: float64

In [85]:
df[['annual_income', 'interaction_count']].corrwith(df.converted)

annual_income        0.078256
interaction_count    0.374573
dtype: float64

### Split the data

- Split your data in train/val/test sets with 60%/20%/20% distribution.
- Use Scikit-Learn for that (the `train_test_split` function) and set the seed to `42`.
- Make sure that the target value `converted` is not in your dataframe.

### Question 3

- Calculate the mutual information score between `converted` and other categorical variables in the dataset. Use the training set only.
- Round the scores to 2 decimals using `round(score, 2)`.

Which of these variables has the biggest mutual information score?

- `industry`
- `location`
- `lead_source`
- `employment_status`

In [86]:
from sklearn.model_selection import train_test_split

In [87]:
df_full_train , df_test = train_test_split(df, test_size=0.2, random_state=42)

In [88]:
len(df_full_train), len(df_test)

(1169, 293)

In [89]:
df_train , df_val = train_test_split(df_full_train, test_size=0.25, random_state=42)

In [90]:
len(df_train), len(df_val), len(df_test)

(876, 293, 293)

In [91]:
df_train.columns

Index(['lead_source', 'industry', 'number_of_courses_viewed', 'annual_income',
       'employment_status', 'location', 'interaction_count', 'lead_score',
       'converted'],
      dtype='object')

In [92]:
# del df_full_train['converted']
del df_test['converted']
# del df_train['converted']
del df_val['converted']

In [93]:
df_train

lead_source       industry  number_of_courses_viewed  annual_income  \
1077        paid_ads         retail                         0        58472.0   
463   organic_search  manufacturing                         3        71738.0   
842         paid_ads     technology                         3        81973.0   
835               NA     technology                         1        74956.0   
837   organic_search         retail                         3        59335.0   
...              ...            ...                       ...            ...   
725   organic_search          other                         1        43907.0   
401     social_media         retail                         3        64969.0   
957               NA      education                         3        89042.0   
992     social_media  manufacturing                         1            NaN   
1267          events        finance                         4        50259.0   

     employment_status       location  interaction_count  lead_score  \
1077           student    middle_east                  5        0.03   
463            student    middle_east                  6        0.77   
842           employed  north_america                  2        0.59   
835           employed         europe                  3        0.34   
837            student      australia                  1        0.98   
...                ...            ...                ...         ...   
725           employed      australia                  4        0.33   
401           employed  north_america                  1        0.18   
957           employed           asia                  4        0.75   
992      self_employed         europe                  1        0.65   
1267           student           asia                  3        0.54   

      converted  
1077          0  
463           1  
842           1  
835           1  
837           1  
...         ...  
725           1  
401           0  
957           1  
992           0  
1267          1  

[876 rows x 9 columns]

In [94]:
df_train = df_train.reset_index(drop=True)

In [95]:
df_train


lead_source       industry  number_of_courses_viewed  annual_income  \
0          paid_ads         retail                         0        58472.0   
1    organic_search  manufacturing                         3        71738.0   
2          paid_ads     technology                         3        81973.0   
3                NA     technology                         1        74956.0   
4    organic_search         retail                         3        59335.0   
..              ...            ...                       ...            ...   
871  organic_search          other                         1        43907.0   
872    social_media         retail                         3        64969.0   
873              NA      education                         3        89042.0   
874    social_media  manufacturing                         1            NaN   
875          events        finance                         4        50259.0   

    employment_status       location  interaction_count  lead_score  converted  
0             student    middle_east                  5        0.03          0  
1             student    middle_east                  6        0.77          1  
2            employed  north_america                  2        0.59          1  
3            employed         europe                  3        0.34          1  
4             student      australia                  1        0.98          1  
..                ...            ...                ...         ...        ...  
871          employed      australia                  4        0.33          1  
872          employed  north_america                  1        0.18          0  
873          employed           asia                  4        0.75          1  
874     self_employed         europe                  1        0.65          0  
875           student           asia                  3        0.54          1  

[876 rows x 9 columns]

In [96]:
from sklearn.metrics import mutual_info_score

In [97]:
categorical = ['industry', 'location', 'lead_source','employment_status' ]

In [98]:
mutual_info_score(df_train.industry, df_train.converted)

0.011574521435657112

In [99]:
mutual_info_score(df_train.location, df_train.converted)

0.004464157884038034

In [100]:
def mutual_info_converted(series):
  return mutual_info_score(series, df_train.converted)

In [103]:
mi = df_train[categorical].apply(mutual_info_converted)
mi.round(2).sort_values(ascending=False)

lead_source          0.04
industry             0.01
employment_status    0.01
location             0.00
dtype: float64

### Question 4

- Now let's train a logistic regression.
- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
  - To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
  - `model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)`
- Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

What accuracy did you get?

- 0.64
- 0.74
- 0.84
- 0.94

### Question 5

- Let's find the least useful feature using the _feature elimination_ technique.
- Train a model using the same features and parameters as in Q4 (without rounding).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

- `'industry'`
- `'employment_status'`
- `'lead_score'`

> **Note**: The difference doesn't have to be positive.

### Question 6

- Now let's train a regularized logistic regression.
- Let's try the following values of the parameter `C`: `[0.01, 0.1, 1, 10, 100]`.
- Train models using all the features as in Q4.
- Calculate the accuracy on the validation dataset and round it to 3 decimal digits.

Which of these `C` leads to the best accuracy on the validation set?

- 0.01
- 0.1
- 1
- 10
- 100

> **Note**: If there are multiple options, select the smallest `C`.
